### PR TITLE
Units payload.

### PR DESCRIPTION
**Pull request by summetdev**
Allows you to drop payloads from all of your controlled units.

preview

Suggested by [Makc#2908](https://github.com/Makc-Carbon)

### CODE REVIEW

**Comment by YellOw139** [comment]
But, uh, what if I want to control where every block goes?

Also, do they all go into a line? Or who is going to guarantee that my megas in a mixed unit formation will drop blocks in specific, predictable spots? I doubt there's anyone who can do that. Hence, your idea is well intentioned but too unpredictable to for me or other players to be able to rely on, which makes it useless IMO.

File: core/src/mindustry/input/InputHandler.java [hunk 1]
```diff
@@ -241,6 +241,11 @@ public static void payloadDropped(Unit unit, float x, float y){
             pay.set(x, y);
             pay.dropLastPayload();
             pay.set(prevx, prevy);
+            pay.controlling().each(u -> {
+                if (u instanceof Payloadc p){
```
Comment (Anuken) [inline core/src/mindustry/input/InputHandler.java:245]:
```suggestion
                if(u instanceof Payloadc p){
```

File: core/src/mindustry/input/InputHandler.java [hunk 2]
```diff
@@ -241,6 +241,11 @@ public static void payloadDropped(Unit unit, float x, float y){
...
+                    p.dropLastPayload();
```
Comment (Anuken) [inline core/src/mindustry/input/InputHandler.java:246]: This won't sync properly. You must use `Call.payloadDropped(...)` so positions are correct clientside.

**Review by Anuken — CHANGES_REQUESTED** [review]

**Comment by YellOw139** [comment]
Maybe if it was enabled for units only... Dropping units all at once would be a useful thing, I agree. But with in blocks I don't see its use.

I'm so mixed on this lol.

File: core/src/mindustry/input/InputHandler.java (hunk 2 above)
Comment (summetdev) [reply core/src/mindustry/input/InputHandler.java]: Fixed. https://github.com/Anuken/Mindustry/pull/3010/commits/12e35b643caeb1d19ffd306a884921da93fcc08a

**Comment by BasedUser** [comment]
is it possible to make controlled units take and drop items in sync with your main unit?

**Comment by summetdev** [comment]
> is it possible to make controlled units take and drop items in sync with your main unit?

Yes

**Comment by summetdev** [comment]
>But with blocks I don't see its use.

image

File: core/src/mindustry/input/InputHandler.java [hunk 3]
```diff
@@ -241,11 +241,7 @@ public static void payloadDropped(Unit unit, float x, float y){
             pay.set(x, y);
             pay.dropLastPayload();
             pay.set(prevx, prevy);
-            pay.controlling().each(u -> {
-                if(u instanceof Payloadc p){
-                    p.dropLastPayload();
-                }
-            });
+            pay.controlling().each(u -> Call.payloadDropped(u, u.x, u.y));
```
Comment (Anuken) [inline core/src/mindustry/input/InputHandler.java:244]: I know the `instanceof` check technically isn't necessary, but by omitting it you'll be sending extra payload drop packets to clients that won't do anything with it. It's best to keep the check there.

Comment (summetdev) [reply core/src/mindustry/input/InputHandler.java]: Fixed. https://github.com/Anuken/Mindustry/pull/3010/commits/801eca5ca3b2ada72884880636c9344d6ea999ea

**Pull request merged**